src/Options/ServerOptions.php

```diff
@@ -92,7 +92,7 @@ public function __construct($options = null)
      *
      * @param  array $options
      */
-    public function setFromArray(array $options)
+    protected function setFromArray(array $options)
```

Comment:
We've done this with @danizord on our project lately: you could create a private constructor, and using a static "fromArray" method that construct the object. This way, all the setter could be removed, which will ensure true immutability.